File: android/src/main/java/com/facebook/flipper/plugins/inspector/descriptors/utils/stethocopies/ReflectionUtil.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.flipper.plugins.inspector.descriptors.utils.stethocopies;

import android.util.Log;
import java.lang.reflect.Field;
import javax.annotation.Nullable;

public final class ReflectionUtil {
  private ReflectionUtil() {}

  @Nullable
  public static Class<?> tryGetClassForName(String className) {
    try {
      return Class.forName(className);
    } catch (ClassNotFoundException e) {
      return null;
    }
  }

  @Nullable
  public static Field tryGetDeclaredField(Class<?> theClass, String fieldName) {
    try {
      return theClass.getDeclaredField(fieldName);
    } catch (NoSuchFieldException e) {
      Log.e(
          ReflectionUtil.class.getCanonicalName(),
          String.format("Could not retrieve %s field from %s", fieldName, theClass),
          e);

      return null;
    }
  }

  @Nullable
  public static Object getFieldValue(Field field, Object target) {
    try {
      return field.get(target);
    } catch (IllegalAccessException e) {
      throw new RuntimeException(e);
    }
  }
}

File: android/src/main/java/com/facebook/flipper/plugins/uidebugger/stetho/ReflectionUtil.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.flipper.plugins.uidebugger.stetho

import java.lang.reflect.Field

object ReflectionUtil {
  fun tryGetClassForName(className: String): Class<*>? {
    return try {
      Class.forName(className)
    } catch (e: ClassNotFoundException) {
      null
    }
  }

  fun tryGetDeclaredField(theClass: Class<*>, fieldName: String): Field? {
    return try {
      theClass.getDeclaredField(fieldName)
    } catch (e: NoSuchFieldException) {
      null
    }
  }

  fun getFieldValue(field: Field, target: Any?): Any? {
    return try {
      field[target]
    } catch (e: IllegalAccessException) {
      throw RuntimeException(e)
    }
  }
}
